exporter/loadbalancingexporter/resolver_k8s.go

```diff
@@ -0,0 +1,301 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package loadbalancingexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+var _ resolver = (*k8sResolver)(nil)
+
+var (
+	errNoSvc                        = errors.New("no service specified to resolve the backends")
+	errNoSvcPorts                   = errors.New("no ports specified to resolve the backends")
+	k8sResolverMutator              = tag.Upsert(tag.MustNewKey("resolver"), "k8s service")
+	k8sResolverSuccessTrueMutators  = []tag.Mutator{resolverMutator, successTrueMutator}
+	k8sResolverSuccessFalseMutators = []tag.Mutator{resolverMutator, successFalseMutator}
+)
+
+type k8sResolver struct {
+	logger    *zap.Logger
+	service   string
+	namespace string
+	port      []int32
+
+	handler        *handler
+	once           *sync.Once
+	epsListWatcher cache.ListerWatcher
+	endpointsStore *sync.Map
+
+	endpoints         []string
+	onChangeCallbacks []func([]string)
+
+	stopCh             chan (struct{})
+	updateLock         sync.Mutex
+	shutdownWg         sync.WaitGroup
+	changeCallbackLock sync.RWMutex
+}
+
+func newK8sResolver(clt kubernetes.Interface,
+	logger *zap.Logger,
+	service string,
+	ports []int32) (*k8sResolver, error) {
+
+	if len(service) == 0 {
+		return nil, errNoSvc
+	}
+	if len(ports) == 0 {
+		return nil, errNoSvcPorts
+	}
+
+	nAddr := strings.SplitN(service, ".", 2)
+	name, namespace := nAddr[0], "default"
+	if len(nAddr) > 1 {
+		namespace = nAddr[1]
+	}
+
+	epsSelector := fmt.Sprintf("metadata.name=%s", name)
+	epsListWatcher := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = epsSelector
+			return clt.CoreV1().Endpoints(namespace).List(context.TODO(), options)
```

Comment:
Would it respect a timeout if we send it? What happens if we don't specify and it takes, say, 30s to answer?

```suggestion
			return clt.CoreV1().Endpoints(namespace).List(context.WithTimeout(context.Background(), time.Second), options)
```

Reply:
In fact, it would be nice to add the response time to a metric, so that people will know how long each API call is taking.

Reply:
The `TimeoutSeconds` field in the `options` parameter can specify the timeout period. should we use this field to control the timeout? 
it looks like this
```
options.FieldSelector = epsSelector
options.TimeoutSeconds = pointer.Int64(1)
return clt.CoreV1().Endpoints(namespace).List(context.Background(), options)
```